### ISSUE TITLE
RFC: let path resolution in macro expression be lexical

### ISSUE THREAD

**Issue by MakeNowJust** [crystal-lang/crystal#9556]
In usual expression, path (`Type::Const`) resolution is lexical (i.e. only depending on location where it is wrote). However, in macro expression, path resolution rule is complicated and unpredictable.

For example, a macro defined in namespace resolves path from its receiver. This rule is unique, but it is not useful in fact.

```crystal
module Foo
  A = :Foo

  macro mac
    {% p A %}
  end
end

class Bar
  A = :Bar

  include Foo
end

Foo.mac # it shows `:Foo` here
Bar.mac # and shows `:Bar` here (?)
```

Besides, path in macro hook is resolved from a type which is the reason of invoking this hook.

I think, there are no reason why they are so on. It is unreasonable really. Thus, it is better that path resolutions in any macro expressions unify to resolving as lexical. At least it is the same rule of path resolution in usual expression, so we can accept it easily.

One of benefits of this change is that it makes macros more robust. In other words this prevents either `include` and inheritance breaks a macro. For instance, when a macro refers constant and including type redefines the same name constant, then the macro will be broken currently.

Remaining problem is `Path#resolve` (and similar methods) resolution. Resolution rule in macro expression is diverted to `Path#resolve`. But. if this RFC is accepted, this behavior seems bad obviously. It should be resolved from wrote location.

```crystal
class Foo
  A = :Foo

  macro mac(path)
    {% p path.resolve %}
  end
end

A = :global
Foo.mac(A) # currently it shows `:Foo`, but `:global` is expected.
```

However, in some cases (e.g. DSL), resolving path from macro location is preferred. For this case, we should add overload `Path#resolve(scope : TypeNode)`. It is more explicit than now.

**Comment by asterite** [comment]
Yeah, I think type resolution like it is in methods makes the most sense because, as you say, it can't be broken by inheritance.

Then Path.resolve should resolve the type depending on where the constant is defined. But that's not possible right now because constants don't store the location where they are defined. I think this is the same problem as JSON::Serialize where the converter doesn't work if it's relative to the place where it's declared.
